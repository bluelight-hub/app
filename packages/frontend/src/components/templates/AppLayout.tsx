import React, { useEffect, useState } from 'react';
import { Outlet, useLocation, useNavigate } from 'react-router';
import { findRouteTitle, mainNavigation, adminNavigation } from '../../config/navigation';
import useMediaQuery from '../../hooks/useMediaQuery';
import { useAuth } from '../../hooks/useAuth';
import MobileHeader from '../organisms/MobileHeader';
import Sidebar from '../organisms/Sidebar';

/**
 * AppLayout template - Defines the main layout structure for the app section
 */
const AppLayout: React.FC<{ children?: React.ReactNode }> = () => {
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);
  const location = useLocation();
  const title = findRouteTitle(location.pathname);
  const isMobile = useMediaQuery('(max-width: 1024px)'); // lg breakpoint in Tailwind
  const _navigate = useNavigate();
  const { isAdmin } = useAuth();

  // Entfernt - die Navigation wird jetzt durch den EinsatzGuard gehandhabt

  // Schließt die Sidebar automatisch, wenn der Viewport auf mobile Größe wechselt
  useEffect(() => {
    if (isMobile) {
      setIsMobileMenuOpen(false);
    }
  }, [isMobile]);

  // Combine main and admin navigation for admin users
  const navigation = isAdmin() ? [...mainNavigation, ...adminNavigation] : mainNavigation;

  return (
    <div className="min-h-screen bg-white dark:bg-gray-900 text-black dark:text-white">
      {/* Unified Sidebar - entscheidet intern zwischen Mobile und Desktop */}
      <Sidebar
        isOpen={isMobileMenuOpen}
        onClose={() => setIsMobileMenuOpen(false)}
        isMobile={isMobile}
        navigation={navigation}
      />

      {/* Mobile Header */}
      <MobileHeader title={title} onOpenSidebar={() => setIsMobileMenuOpen(true)} />

      {/* Main Content */}
      <main className="py-10 lg:pl-72">
        <div className="px-4 sm:px-6 lg:px-8">
          <Outlet />
        </div>
      </main>
    </div>
  );
};

export default AppLayout;
