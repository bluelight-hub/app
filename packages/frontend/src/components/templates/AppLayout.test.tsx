import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { vi } from 'vitest';

// Mock the entire AppLayout component
vi.mock('./AppLayout', () => ({
    default: function MockAppLayout({ children }: { children?: React.ReactNode }) {
        const [isMobileMenuOpen, setIsMobileMenuOpen] = React.useState(false);

        return (
            <div data-testid="app-layout" className="min-h-screen bg-white dark:bg-gray-900">
                <div
                    data-testid="sidebar"
                    data-is-open={isMobileMenuOpen}
                    data-nav-items="1"
                >
                    <button data-testid="close-sidebar-button" onClick={() => setIsMobileMenuOpen(false)}>
                        Close Sidebar
                    </button>
                </div>

                <header data-testid="mobile-header">
                    <span data-testid="page-title">Dashboard</span>
                    <button data-testid="open-sidebar-button" onClick={() => setIsMobileMenuOpen(true)}>
                        Open Sidebar
                    </button>
                </header>

                <main className="py-10 lg:pl-72">
                    <div className="px-4 sm:px-6 lg:px-8">
                        {children || <div data-testid="outlet-content">Page Content</div>}
                    </div>
                </main>
            </div>
        );
    }
}));

// Import the component after mocking it
import AppLayout from './AppLayout';

// Mock dependencies
vi.mock('../../config/navigation', () => ({
    findRouteTitle: vi.fn((path) => path === '/dashboard' ? 'Dashboard' : 'Unknown'),
    mainNavigation: [
        { name: 'Dashboard', href: '/dashboard', icon: 'DashboardIcon' }
    ]
}));

describe('AppLayout', () => {
    const renderComponent = () => {
        return render(
            <AppLayout />
        );
    };

    it('renders correctly with all components', () => {
        renderComponent();

        // Check if all main components are rendered
        expect(screen.getByTestId('app-layout')).toBeInTheDocument();
        expect(screen.getByTestId('sidebar')).toBeInTheDocument();
        expect(screen.getByTestId('mobile-header')).toBeInTheDocument();
        expect(screen.getByTestId('outlet-content')).toBeInTheDocument();
    });

    it('displays correct title in the header', () => {
        renderComponent();

        // Check if the correct title is displayed in the header
        expect(screen.getByTestId('page-title').textContent).toBe('Dashboard');
    });

    it('toggles mobile menu when sidebar button is clicked', () => {
        renderComponent();

        // Initially sidebar should be closed
        expect(screen.getByTestId('sidebar').getAttribute('data-is-open')).toBe('false');

        // Open sidebar
        fireEvent.click(screen.getByTestId('open-sidebar-button'));
        expect(screen.getByTestId('sidebar').getAttribute('data-is-open')).toBe('true');

        // Close sidebar
        fireEvent.click(screen.getByTestId('close-sidebar-button'));
        expect(screen.getByTestId('sidebar').getAttribute('data-is-open')).toBe('false');
    });

    it('has correct layout classes', () => {
        renderComponent();

        // Check correct classes for layout
        const layout = screen.getByTestId('app-layout');
        expect(layout).toHaveClass('min-h-screen');
        expect(layout).toHaveClass('bg-white');
    });

    // Snapshot test
    it('matches snapshot', () => {
        const { container } = renderComponent();
        expect(container).toMatchSnapshot();
    });
}); 