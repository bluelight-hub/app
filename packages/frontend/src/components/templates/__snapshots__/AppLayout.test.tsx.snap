// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppLayout > matches snapshot 1`] = `
<div>
  <div
    class="min-h-screen bg-white dark:bg-gray-900"
    data-testid="app-layout"
  >
    <div
      data-is-mobile="false"
      data-is-open="false"
      data-nav-items="1"
      data-testid="sidebar"
    >
      <button
        data-testid="close-sidebar-button"
      >
        Close Sidebar
      </button>
    </div>
    <header
      data-testid="mobile-header"
    >
      <span
        data-testid="page-title"
      >
        Dashboard
      </span>
      <button
        data-testid="open-sidebar-button"
      >
        Open Sidebar
      </button>
    </header>
    <main
      class="py-10 lg:pl-72"
    >
      <div
        class="px-4 sm:px-6 lg:px-8"
      >
        <div
          data-testid="outlet-content"
        >
          Page Content
        </div>
      </div>
    </main>
  </div>
</div>
`;
